src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Debugger_InProc.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using EnvDTE;
+
+namespace Microsoft.VisualStudio.IntegrationTest.Utilities.InProcess
+{
+    internal class Debugger_InProc : InProcComponent
+    {
+        public static Debugger_InProc Create() => new Debugger_InProc();
+
+        public void StartDebugging(bool waitForBreakMode)
+        {
+            var debuggerEventsMonitor = new VsDebuggerEventsMonitor(GetDTE().Events.DebuggerEvents);
+            if (waitForBreakMode)
+            {
+                StartAndWaitForDebuggerBreakMode(debuggerEventsMonitor);
+            }
+            else
+            {
+                StartAndWaitForDebuggerRunMode(debuggerEventsMonitor);
+            }
+        }
+
+        public void Continue(bool waitForBreakMode)
+        {
+            // Sometimes you will get a COM exception if you attempt to continue too quickly
+            RetryHelper.Try(() => GetDTE().Debugger.Go(waitForBreakMode));
+        }
+
+        private void StartAndWaitForDebuggerDesignMode(VsDebuggerEventsMonitor debuggerEventsMonitor)
+        {
+            using (var semaphore = new SemaphoreSlim(1))
```

Comment:
You could also consider something like a ManualResetEvent which might be closer to your "this is a gate to let something through" pattern.